crates/kubelet/src/fs_watch/mod.rs

```diff
@@ -0,0 +1,315 @@
+                    );
+                    HashSet::new()
+                }
+            };
+
+            loop {
+                let current_paths: HashSet<PathBuf> = match get_dir_list(&path).await {
+                    Ok(set) => set,
+                    Err(e) => {
+                        error!(
+                            "Unable to refresh directory {}, will attempt again: {:?}",
+                            path.display(),
+                            e
+                        );
+                        if let Err(e) = tx.send(Err(NotifyError::io(e))) {
+                            error!("Unable to send error {:?} due to channel being closed", e.0);
+                        }
+                        continue;
+                    }
+                };
+
+                // Do a difference between cached and current paths (current - cached) to detect set of creates
+                handle_creates(tx.clone(), current_paths.difference(&path_cache).cloned());
+
+                // Do a difference between cached and current paths (cached - current) to detect set of deletes
+                handle_deletes(tx.clone(), path_cache.difference(&current_paths).cloned());
+
+                // Now we can set current to cached
+                path_cache = current_paths;
+
+                time::delay_for(Duration::from_secs(WAIT_TIME)).await;
+            }
+        });
+        rx
+    }
+
+    async fn get_dir_list(path: &PathBuf) -> Result<HashSet<PathBuf>, std::io::Error> {
+        // What does this monstrosity do? Well, due to async and all the random streaming involved
+        // this:
+        // 1. Reads the directory as a stream
+        // 2. Maps the stream to a Vec of entries and handles any errors
+        // 3. Converts the entries to PathBufs and puts them in a HashSet
+        tokio::fs::read_dir(path)
+            .await?
+            .collect::<Result<Vec<DirEntry>, _>>()
+            .await
+            .map(|entries| {
+                entries
+                    .into_iter()
+                    .map(|e| e.path())
+                    .collect::<HashSet<PathBuf>>()
+            })
+    }
+
+    fn handle_creates(
+        tx: UnboundedSender<NotifyResult<Event>>,
+        items: impl Iterator<Item = PathBuf>,
+    ) {
+        let paths: Vec<PathBuf> = items.collect();
+        // If there were no paths, it means there weren't any new files, so return
+        if paths.is_empty() {
+            return;
+        }
+        let event = Event {
+            kind: EventKind::Create(CreateKind::Any),
+            paths,
+            ..Default::default()
+        };
+        if let Err(e) = tx.send(Ok(event)) {
+            // At this point there isn't much we can do as the channel is closed. So just log an
+            // error
+            error!(
+                "Unable to send event {:?} due to the channel being closed",
+                e.0
+            );
+        }
+    }
+
+    fn handle_deletes(
```

Comment:
This implementation seems almost identical to `handle_creates`.  Consider pulling out a `handle_changes` (or `send_changes`) that takes an `EventKind` as a parameter and putting all the common logic there.